src/man/kiwix-serve.1
.TH KIWIX 1 "10 July 2023"

.SH NAME
kiwix-serve \- Kiwix HTTP Server

.SH SYNOPSIS

.B kiwix-serve --library [OPTIONS] LIBRARY_FILE_PATH
.br
.B kiwix-serve [OPTIONS] ZIM_FILE_PATH ...

.SH DESCRIPTION
The \fBkiwix-serve\fR command is used to run a stand-alone HTTP server for serving ZIM contents over the network.

.SH ARGUMENTS
.TP
\fBLIBRARY_FILE_PATH\fR
Path of a library file (XML or OPDS) listing ZIM files to serve. To be used only with the --library option. Multiple library files can be provided as a semicolon (;) separated list.

.TP
\fBZIM_FILE_PATH ...\fR
ZIM file path(s). Multiple arguments are allowed.

.SH OPTIONS
.TP
\fB--library\fR
By default, kiwix-serve expects a list of ZIM files as command line arguments. Providing the --library option tells kiwix-serve that the command line argument is rather a library file (XML or OPDS).

.TP
\fB-i ADDR, --address=ADDR\fR
Listen only on this IP address. By default, the server listens on all available IP addresses. Alternatively, you can use special values to define which types of connections to accept:

all : Listen for connections on all IP addresses (IPv4 and IPv6).
.br
ipv4 : Listen for connections on all IPv4 addresses.
.br
ipv6 : Listen for connections on all IPv6 addresses.

.TP
\fB-p PORT, --port=PORT\fR
TCP port on which to listen for HTTP requests (default: 80).

.TP
\fB-r ROOT, --urlRootLocation=ROOT\fR
URL prefix on which the content should be made available (default: empty).

.TP
\fB-d, --daemon\fR
Detach the HTTP server daemon from the main process.

.TP
\fB-a PID, --attachToProcess=PID\fR
Exit when the process with id PID stops running.

.TP
\fB-M, --monitorLibrary\fR
Monitor the library file and reload it automatically when it changes.

Library reloading can be forced anytime by sending a SIGHUP signal to the
\*(lqkiwix-serve\*(rq process (this works regardless of the presence of the
\*(lq--monitorLibrary\*(rq/\*(lq-M\*(rq option).

.TP
\fB-m, --nolibrarybutton\fR
Disable the library home button in the ZIM viewer toolbar.

.TP
\fB-n, --nosearchbar\fR
Disable the search box in the ZIM viewer toolbar.

.TP
\fB-b, --blockexternal\fR
Prevent users from directly navigating to external resources via links in ZIM content.

.TP
\fB-t N, --threads=N\fR
Number of threads to run in parallel (default: 4).

.TP
\fB-s N, --searchLimit=N\fR
Maximum number of ZIM files in a fulltext multizim search (default: No limit).

.TP
\fB-z, --nodatealiases\fR
Create URL aliases for each content by removing the date embedded in the file name.

The expected format of the date in the filename is \*(lq_YYYY-MM\*(rq. For example, a ZIM file named \*(lqwikipedia_en_all_2020-08.zim\*(rq will be accessible both as \*(lqwikipedia_en_all_2020-08\*(rq and \*(lqwikipedia_en_all\*(rq.

.TP
\fB-c PATH, --customIndex=PATH\fR
Override the welcome page with a custom HTML file.

.TP
\fB-L N, --ipConnectionLimit=N\fR
Max number of (concurrent) connections per IP (default: infinite, recommended: >= 6).

.TP
\fB-k, --skipInvalid\fR
Startup even when ZIM files are invalid (those will be skipped)

.TP
\fB-v, --verbose\fR
Print debug log to STDOUT.

.TP
\fB-V, --version\fR
Print the software version.

.TP
\fB-h, --help\fR
Print a help message.

.SH EXAMPLES
Serve a single ZIM file:
.sp
.nf
.B kiwix-serve myzim.zim
.fi

Serve multiple ZIM files:
.sp
.nf
.B kiwix-serve zim1.zim zim2.zim zim3.zim
.fi

Serve ZIM files from a library:
.sp
.nf
.B kiwix-serve --library library.xml
.fi

.SH DOCUMENTATION
Online documentation: https://kiwix-tools.readthedocs.io/en/latest/kiwix-serve.html
.br
Source code: https://github.com/kiwix/kiwix-tools
.br
More info: https://wiki.kiwix.org/wiki/Kiwix-serve

.SH AUTHORS
Emmanuel Engelhart <kelson@kiwix.org>
.br
Vasudev Kamath <kamathvasudev@gmail.com>


Changelog
kiwix-tools 3.8.2
=================

 * Other
   - Fix PPA publishing (@kelson42 #800)
   - Use windows-2025 CI runner (@kelson42 #805)
   - Updated documentation of lang and category filters in catalog query API endpoints (@veloman-yunkan #804)

kiwix-tools 3.8.1
=================

 * Kiwix server
   - Hide port number in URL when server is running on port 80 (@vighnesh-sawant #763)
   - Better deal with container /data dir permissions (@kelson42 #787)

 * Other
   - Fix kiwix-manage docopt integration (@kelson42 #783)

kiwix-tools 3.8.0
=================

 * Kiwix server
   - Improve message when server is running on standard port (@vighnesh-sawant #763)
   - Update container base image to latest Alpline version (@yashgoyal0110 #771)
   - Container image to use unprivileged user (@Sedetisu #755)
   - Empty urlRootLocation doesn't disable book preview links anymore (@veloman-yunkan #1224)
   - Add support of IPv6 (@sgourdas #673 #704)
   - Popups are allowed to escape the browser sandbox (@veloman-yunkan #1208)

 * Other
   - Stop publishing on Ubuntu 20.04 PPA (@kelson42 #748)
   - Use Docopt for command line argument parsing (@mgautierfr #695)

 * Compilation & Packaging
   - Based on libkiwix 14.1.0 (@kelson42 #773)
   - Improve CI/CD (@kelson42 #681 #698 #702 #705 #720 #722, @mgautierfr #695)

kiwix-tools 3.7.0
=================

 * Fixed ZIM name vs Book name confusion in documentation (@veloman-yunkan #663)
 * Fixes compilation dependencies to rely on appropriate version (@kelson42 #667)
 * New --skipInvalid Kiwix Server command line option (@schuellerf @kelson42 #666)

kiwix-tools 3.6.0
=================

 * Improved kiwix-serve man page (@iArchitSharma #626)
 * C++17 compliant code base (@mgautierfr #636)
 * Support of libkiwix13 (@mgautierfr #633)
 * Additional docker images archs for armv6 and i386 (@rgaudin #622)

kiwix-tools 3.5.0
=================

 * Do not use `--static` option when compiling on MacOs (@mgautierfr #615)
 * Move main branch from `master` to `main`.
 * Fix docker image (@jacroe #597)
 * Various CI improvements (@kelson42)

kiwix-serve
-----------

 * Add documentation about the kiwix-serve API (@veloman-yunkan #586)
   https://kiwix-tools.readthedocs.io/en/latest/kiwix-serve.html#http-api

kiwix-tools 3.4.0
=================

 * Remove last reference to kiwix-read tool (@legoktm #569)

kiwix-serve
-----------

 * Fix broken indentation in usage (@kelson42 #560)
 * Exit if wrong arguments are passed (@kelson42 #567)
 * Do not allow multiple values for same option (@juuz0 #564)
 * Fix default location of "rootLocation" (@rgaudin #571)
 * [DOCKER] Change default port to 8080 (@neyder #581)
 * [DOCKER] Simplify dockerfile (@rgaudin #582)

kiwix-manage
------------

 * Fix man page (@kelson42 #576)

kiwix-tools 3.3.0
=================

 * Remove kiwix-read tool (@veloman-yunkan #535)

kiwix-serve
-----------

 * Add an option to limit the number of connections for a same IP (@juuz0 #534)
 * Add an option to limit the number of zim in a multizim fulltext search (@mgautierfr #558)

kiwix-search
------------

 * Remove usage of libkiwix's deprecated api (@veloman-yunkan #535)

kiwix-manage
------------

 * Correctly return a value !0 if something went wrong (@mgautierfr #553)


kiwix-tools 3.2.0
=================

 * Print the version of all dependencies (@kelson42 #516)
 * Better Docker images (@kelson42 @rgaudin)
 * Update Readme (@kelson42)
 * Build debian packages on CI (@legoktm #394)
 * Add man pages for kiwix-read and kiwix-search (@legoktm #392)
 * Various fixes (@legoktm @hashworks @mgautierfr)


kiwix-serve
-----------

 * Print the url on which a user can connect to on startup (@juuz0 #499 #522)
 * Reload library on SIGHUP signal (@veloman-yunkan #497)
 * Add a option `--monitorLibrary` to monitor and automically reload the library
   (@veloman-yunkan #503)
 * Correct handling of SIGTERM and SIGINT (@veloman-yunkan #488)
 * Add `--customIndexTemplate` option (@manan #477)
 * Add `--help` option (@kelson42 #511)


kiwix-tools 3.1.2
=================

 * Use new threadsafe API of kiwix-lib to do suggestions search.

kiwix-tools 3.1.1
=================

 * Fix compilation on Windows' CI

kiwix-tools 3.1.0
=================

 * [SERVER] Add option to block external links

kiwix-tools 3.0.3
=================

 * [MANAGER] Fix broken --version argument parsing

kiwix-tools 3.0.2
=================

 * New option --version for all tools
 * Remove benchmark.sh file.
 * [DOCKER] Add ability to download a file at container start.
 * [CI] Move to github actions instead of travis.
 * [SERVER] Trust the given library by default.
 * [SERVER] Add shortcut alias for option `--address` and `--nodatealias`

kiwix-tools 3.0.1
=================

 * Fix --nodatealiases inverted logic regression

kiwix-tools 3.0.0
=================

 * Move kiwix-serve implementation in kiwix-lib.

kiwix-tools 2.1.0
=================

 * Fix few compilation errors.

kiwix-serve
-----------

 * Use new api to filter the library.
 * Mobile friendly top bar.
 * Add notag parameter to be able to exclude tags from the zim search.


kiwix-tools 2.0.0
=================

kiwix-manage
-----------

 * Better usage()
 * Adding multiple files bug fix
 * Remove download command.

kiwix-serve
-----------

 * Better usage()
 * Display properly welcome page on 3 columns
 * New welcome page footer "Powered by Kiwix"

kiwix-tools 1.2.1
=================

kiwix-serve
-----------

 * Always use POLL when avaible.

kiwix-tools 1.2.0
=================

 * Remove rpath for installed binaries.

kiwix-serve
-----------

 * New Dockerfile of kiwix-serve
 * New --nodatealiases option
 * Do not use POLL on windows

kiwix-manage
------------

 * Do not show all books if book ids has been provided.
 * Be able to add several zim files in the same time in a library.

kiwix-tools 1.1.0
=================

kiwix-serve
-----------

 * Fix bug about handling of absolute url in old zim file.

src/server/kiwix-serve.cpp

/*
 * Copyright 2009-2019 Emmanuel Engelhart <kelson@kiwix.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU  General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 */

#include <docopt/docopt.h>
#include <kiwix/manager.h>
#include <kiwix/server.h>
#include <kiwix/name_mapper.h>
#include <kiwix/tools.h>

#ifdef _WIN32
# include <windows.h>
#else
# include <unistd.h>
# include <signal.h>
#endif
#include <sys/stat.h>

#ifdef __APPLE__
# import <sys/sysctl.h>
# import <sys/types.h>
# define MIBSIZE 4
#endif

#include "../version.h"

#define DEFAULT_THREADS 4
#define LITERAL_AS_STR(A) #A
#define AS_STR(A) LITERAL_AS_STR(A)


static const char USAGE[] =
R"(Deliver ZIM file(s) articles via HTTP

Usage:
 kiwix-serve [options] ZIMPATH ...
 kiwix-serve [options] (-l | --library) LIBRARYPATH
 kiwix-serve -h | --help
 kiwix-serve -V | --version

Mandatory arguments:
  LIBRARYPATH  Library file path (XML or OPDS) listing ZIM file to serve. To be used only with the --library argument."
  ZIMPATH      ZIM file path(s)

Options:
 -h --help                               Print this help
 -a <pid> --attachToProcess=<pid>        Exit if given process id is not running anymore [default: 0]
 --catalogOnly                           Serve only the library catalog
 --contentServerURL=<url>                Root URL of the server serving ZIM content for this library
 -d --daemon                             Detach the HTTP server daemon from the main process
 -i <address> --address=<address>        Listen only on the specified IP address. Specify 'ipv4', 'ipv6' or 'all' to listen on all IPv4, IPv6 or both types of addresses, respectively [default: all]
 -M --monitorLibrary                     Monitor the library file and reload it automatically
 -m --nolibrarybutton                    Don't print the builtin home button in the builtin top bar overlay
 -n --nosearchbar                        Don't print the builtin bar overlay on the top of each served page
 -b --blockexternal                      Prevent users from directly accessing external links
 -p <port> --port=<port>                 Port on which to listen to HTTP requests [default: 80]
 -r <root> --urlRootLocation=<root>      URL prefix on which the content should be made available [default: /]
 -s <limit> --searchLimit=<limit>        Maximun number of zim in a fulltext multizim search [default: 0]
 -t <threads> --threads=<threads>        Number of threads to run in parallel [default: )" AS_STR(DEFAULT_THREADS) R"(]
 -v --verbose                            Print debug log to STDOUT
 -V --version                            Print software version
 -z --nodatealiases                      Create URL aliases for each content by removing the date
 -c <path> --customIndex=<path>          Add path to custom index.html for welcome page
 -L <limit> --ipConnectionLimit=<limit>  Max number of (concurrent) connections per IP [default: 0] (recommended: >= 6)
 -k --skipInvalid                        Startup even when ZIM files are invalid (those will be skipped)

Documentation:
  Source code   https://github.com/kiwix/kiwix-tools
  More info     https://wiki.kiwix.org/wiki/Kiwix-serve
                https://kiwix-tools.readthedocs.io/en/latest/kiwix-serve.html
)";

std::string loadCustomTemplate (std::string customIndexPath) {
  customIndexPath = kiwix::isRelativePath(customIndexPath) ?
                      kiwix::computeAbsolutePath(kiwix::getCurrentDirectory(), customIndexPath) :
                      customIndexPath;
  if (!kiwix::fileReadable(customIndexPath)) {
    throw std::runtime_error("No such file exist (or file is not readable) " + customIndexPath);
  }
  if (kiwix::getMimeTypeForFile(customIndexPath) != "text/html") {
    throw std::runtime_error("Invalid File Mime Type " + kiwix::getMimeTypeForFile(customIndexPath));
  }
  std::string indexTemplateString = kiwix::getFileContent(customIndexPath);

  if (indexTemplateString.empty()) {
    throw std::runtime_error("Unreadable or empty file " + customIndexPath);
  }
  return indexTemplateString;
}

#ifndef _WIN32
volatile sig_atomic_t waiting = false;
volatile sig_atomic_t libraryMustBeReloaded = false;
void handle_sigterm(int signum)
{
    if ( waiting == false ) {
        _exit(signum);
    }
    waiting = false;
}

void handle_sighup(int signum)
{
  libraryMustBeReloaded = true;
}

typedef void (*SignalHandler)(int);

void set_signal_handler(int sig, SignalHandler handler)
{
    struct sigaction sa;
    sigaction(sig, NULL, &sa);
    sa.sa_handler = handler;
    sigaction(sig, &sa, NULL);
}

void setup_sighandlers()
{
    set_signal_handler(SIGTERM, &handle_sigterm);
    set_signal_handler(SIGINT,  &handle_sigterm);
    set_signal_handler(SIGHUP,  &handle_sighup);
}
#else
bool waiting = false;
bool libraryMustBeReloaded = false;
#endif

uint64_t fileModificationTime(const std::string& path)
{
#if defined(_WIN32) && !defined(stat)
#define stat _stat
#endif
  struct stat fileStatData;
  if ( stat(path.c_str(), &fileStatData) == 0 ) {
    return fileStatData.st_mtime;
  }
  return 0;
#ifdef _WIN32
#undef stat
#endif
}

uint64_t newestFileTimestamp(const std::vector<std::string>& paths)
{
  uint64_t t = 0;
  for ( const auto& p : paths ) {
    t = std::max(t, fileModificationTime(p));
  }

  return t;
}

bool reloadLibrary(kiwix::Manager& mgr, const std::vector<std::string>& paths)
{
    try {
      std::cout << "Loading the library from the following files:\n";
      for ( const auto& p : paths ) {
        std::cout << "\t" << p << std::endl;
      }
      mgr.reload(paths);
      std::cout << "The library was successfully loaded." << std::endl;
      return true;
    } catch ( const std::runtime_error& err ) {
      std::cerr << "ERROR: " << err.what() << std::endl;
      std::cerr << "Errors encountered while loading the library." << std::endl;
      return false;
    }
}

// docopt::value::isLong() is counting repeated values.
// It doesn't check if the string can be parsed as long.
// (Contrarly to `asLong` which will try to convert string to long)
// See https://github.com/docopt/docopt.cpp/issues/62
// `isLong` is a small helper to get if the value can be parsed as long.
inline bool isLong(const docopt::value& v) {
  try {
    v.asLong();
    return true;
  } catch (...) {
    return false;
  }
}

#define FLAG(NAME, VAR) if (arg.first == NAME) { VAR = arg.second.asBool(); continue; }
#define STRING(NAME, VAR) if (arg.first == NAME && arg.second.isString() ) { VAR = arg.second.asString(); continue; }
#define STRING_LIST(NAME, VAR, ERRORSTR) if (arg.first == NAME) { if (arg.second.isStringList()) { VAR = arg.second.asStringList(); continue; } else { errorString = ERRORSTR; break; } }
#define INT(NAME, VAR, ERRORSTR) if (arg.first == NAME ) { if (isLong(arg.second)) { VAR = arg.second.asLong(); continue; } else { errorString = ERRORSTR; break; } }

// Older version of docopt doesn't declare Options. Let's declare it ourself.
using Options = std::map<std::string, docopt::value>;

int main(int argc, char** argv)
{
#ifndef _WIN32
  setup_sighandlers();
#endif

  std::string rootLocation = "/";
  auto library = kiwix::Library::create();
  unsigned int nb_threads = DEFAULT_THREADS;
  std::vector<std::string> zimPathes;
  std::string libraryPath;
  std::string rootPath;
  std::string address;
  std::string customIndexPath="";
  std::string indexTemplateString="";
  int serverPort = 80;
  bool catalogOnlyFlag = false;
  std::string contentServerURL;
  bool daemonFlag [[gnu::unused]] = false;
  bool helpFlag = false;
  bool noLibraryButtonFlag = false;
  bool noSearchBarFlag = false;
  bool noDateAliasesFlag = false;
  bool blockExternalLinks = false;
  bool isVerboseFlag = false;
  bool monitorLibrary = false;
  bool versionFlag = false;
  unsigned int PPID = 0;
  int ipConnectionLimit = 0;
  int searchLimit = 0;
  bool skipInvalid = false;

  std::string errorString;

  Options args;
  try {
    args = docopt::docopt_parse(USAGE, {argv+1, argv+argc}, false, false);
  } catch (docopt::DocoptArgumentError const & error) {
    std::cerr << error.what() << std::endl;
    std::cerr << USAGE << std::endl;
    return -1;
  }

  for (auto const& arg: args) {
    FLAG("--help", helpFlag)
    FLAG("--catalogOnly", catalogOnlyFlag)
    STRING("--contentServerURL", contentServerURL)
    FLAG("--daemon", daemonFlag)
    FLAG("--verbose", isVerboseFlag)
    FLAG("--nosearchbar", noSearchBarFlag)
    FLAG("--blockexternal", blockExternalLinks)
    FLAG("--nodatealiases", noDateAliasesFlag)
    FLAG("--nolibrarybutton",noLibraryButtonFlag)
    FLAG("--monitorLibrary", monitorLibrary)
    FLAG("--skipInvalid", skipInvalid)
    FLAG("--version", versionFlag)
    STRING("LIBRARYPATH", libraryPath)
    INT("--port", serverPort, "Port must be an integer")
    INT("--attachToProcess", PPID, "Process to attach must be an integer")
    STRING("--address", address)
    INT("--threads", nb_threads, "Number of threads must be an integer")
    STRING("--urlRootLocation", rootLocation)
    STRING("--customIndex", customIndexPath)
    INT("--ipConnectionLimit", ipConnectionLimit, "IP connection limit must be an integer")
    INT("--searchLimit", searchLimit, "Search limit must be an integer")
    STRING_LIST("ZIMPATH", zimPathes, "ZIMPATH must be a string list")
 }

 if (!errorString.empty()) {
   std::cerr << errorString << std::endl;
   std::cerr << USAGE << std::endl;
   return -1;
 }

 if (helpFlag) {
   std::cout << USAGE << std::endl;
   return 0;
 }

 if (versionFlag) {
   version();
   return 0;
 }

  /* Setup the library manager and get the list of books */
  kiwix::Manager manager(library);
  std::vector<std::string> libraryPaths;
  if (!libraryPath.empty()) {
    libraryPaths = kiwix::split(libraryPath, ";");
    if ( !reloadLibrary(manager, libraryPaths) ) {
      exit(1);
    }

    /* Check if the library is not empty (or only remote books)*/
    if (library->getBookCount(true, false) == 0) {
      std::cerr << "The XML library file '" << libraryPath
           << "' is empty (or has only remote books)." << std::endl;
    }
  } else {
    std::vector<std::string>::iterator it;
    for (it = zimPathes.begin(); it != zimPathes.end(); it++) {
      if (!manager.addBookFromPath(*it, *it, "", false)) {
        if (skipInvalid) {
          std::cerr << "Skipping invalid '" << *it << "' ...continuing" << std::endl;
        } else {
          std::cerr << "Unable to add the ZIM file '" << *it
               << "' to the internal library." << std::endl;
          exit(1);
        }
      }
    }
  }
  auto libraryFileTimestamp = newestFileTimestamp(libraryPaths);
  auto curLibraryFileTimestamp = libraryFileTimestamp;

  kiwix::IpMode ipMode = kiwix::IpMode::AUTO;

  if (address == "all") {
    address.clear();
    ipMode = kiwix::IpMode::ALL;
  } else if (address == "ipv4") {
    address.clear();
    ipMode = kiwix::IpMode::IPV4;
  } else if (address == "ipv6") {
    address.clear();
    ipMode = kiwix::IpMode::IPV6;
  }

#ifndef _WIN32
  /* Fork if necessary */
  if (daemonFlag) {
    pid_t pid;

    /* Fork off the parent process */
    pid = fork();
    if (pid < 0) {
      exit(1);
    }

    /* If we got a good PID, then
       we can exit the parent process. */
    if (pid > 0) {
      exit(0);
    }
  }
#endif

  auto nameMapper = std::make_shared<kiwix::UpdatableNameMapper>(library, noDateAliasesFlag);
  kiwix::Server server(library, nameMapper);

  if (!customIndexPath.empty()) {
    try {
      indexTemplateString = loadCustomTemplate(customIndexPath);
    } catch (std::runtime_error& e) {
      std::cerr << "ERROR: " << e.what() << std::endl;
      exit(1);
    }
  }

  server.setAddress(address);
  server.setRoot(rootLocation);
  server.setPort(serverPort);
  server.setNbThreads(nb_threads);
  server.setVerbose(isVerboseFlag);
  server.setTaskbar(!noSearchBarFlag, !noLibraryButtonFlag);
  server.setBlockExternalLinks(blockExternalLinks);
  server.setIndexTemplateString(indexTemplateString);
  server.setIpConnectionLimit(ipConnectionLimit);
  server.setMultiZimSearchLimit(searchLimit);
  server.setIpMode(ipMode);
  server.setCatalogOnlyMode(catalogOnlyFlag);
  while ( !contentServerURL.empty() && contentServerURL.back() == '/' )
    contentServerURL.pop_back();
  server.setContentServerUrl(contentServerURL);

  if (! server.start()) {
    exit(1);
  }
  
  std::cout << "The Kiwix server is running and can be accessed in the local network at: " << std::endl;
  for (const auto& url : server.getServerAccessUrls()) {
    std::cout << "  - " << url << std::endl;
  }

  /* Run endless (until PPID dies) */
  waiting = true;
  do {
    if (PPID > 0) {
#ifdef _WIN32
      HANDLE process = OpenProcess(SYNCHRONIZE, FALSE, PPID);
      DWORD ret = WaitForSingleObject(process, 0);
      CloseHandle(process);
      if (ret == WAIT_TIMEOUT) {
#elif __APPLE__
      int mib[MIBSIZE];
      struct kinfo_proc kp;
      size_t len = sizeof(kp);

      mib[0] = CTL_KERN;
      mib[1] = KERN_PROC;
      mib[2] = KERN_PROC_PID;
      mib[3] = PPID;

      int ret = sysctl(mib, MIBSIZE, &kp, &len, NULL, 0);
      if (ret != -1 && len > 0) {
#else /* Linux & co */
      std::string procPath = "/proc/" + std::to_string(PPID);
      if (access(procPath.c_str(), F_OK) != -1) {
#endif
      } else {
        waiting = false;
      }
    }

    kiwix::sleep(1000);

    if ( monitorLibrary ) {
      curLibraryFileTimestamp = newestFileTimestamp(libraryPaths);
      if ( !libraryMustBeReloaded ) {
        libraryMustBeReloaded = curLibraryFileTimestamp > libraryFileTimestamp;
      }
    }

    if ( libraryMustBeReloaded && !libraryPaths.empty() ) {
      libraryFileTimestamp = curLibraryFileTimestamp;
      reloadLibrary(manager, libraryPaths);
      nameMapper->update();
      libraryMustBeReloaded = false;
    }
  } while (waiting);

  /* Stop the daemon */
  server.stop();
}
